packages/expo-network/android/src/main/java/expo/modules/network/NetworkModule.kt

```diff
@@ -0,0 +1,158 @@
+package expo.modules.network
+
+import expo.modules.core.Promise
+import expo.modules.core.ExportedModule
+import expo.modules.core.interfaces.ExpoMethod
+import expo.modules.core.interfaces.RegistryLifecycleListener
+
+import android.util.Log
+import android.os.Build
+import android.os.Bundle
+import android.content.Context
+import android.net.NetworkInfo
+import android.net.wifi.WifiInfo
+import android.net.wifi.WifiManager
+import android.net.ConnectivityManager
+import android.net.NetworkCapabilities
+import android.provider.Settings
+
+import java.lang.Exception
+import java.math.BigInteger
+import java.net.InetAddress
+import java.net.UnknownHostException
+import java.nio.ByteOrder
+
+private const val NAME = "ExpoNetwork"
+private val TAG = NetworkModule::class.java.simpleName
+
+class NetworkModule(private val appContext: Context) : ExportedModule(appContext), RegistryLifecycleListener {
+
+  enum class NetworkStateType(val value: String) {
+    NONE("NONE"),
+    UNKNOWN("UNKNOWN"),
+    CELLULAR("CELLULAR"),
+    WIFI("WIFI"),
+    BLUETOOTH("BLUETOOTH"),
+    ETHERNET("ETHERNET"),
+    WIMAX("WIMAX"),
+    VPN("VPN"),
+    OTHER("OTHER");
+
+    val isDefined: Boolean
+      get() = this.value != "NONE" && this.value != "UNKNOWN"
+  }
+
+  override fun getName() = NAME
+
+  private val wifiInfo: WifiInfo
+    get() = try {
+      val manager = appContext.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
+      manager.connectionInfo
+    } catch (e: Exception) {
+      Log.e(TAG, e.message ?: "Wifi information could not be acquired")
+      throw e
+    }
+
+  private fun getConnectionType(netInfo: NetworkInfo?): NetworkStateType = when (netInfo?.type) {
+    ConnectivityManager.TYPE_MOBILE, ConnectivityManager.TYPE_MOBILE_DUN -> NetworkStateType.CELLULAR
+    ConnectivityManager.TYPE_WIFI -> NetworkStateType.WIFI
+    ConnectivityManager.TYPE_BLUETOOTH -> NetworkStateType.BLUETOOTH
+    ConnectivityManager.TYPE_ETHERNET -> NetworkStateType.ETHERNET
+    ConnectivityManager.TYPE_WIMAX -> NetworkStateType.WIMAX
+    ConnectivityManager.TYPE_VPN -> NetworkStateType.VPN
+    else -> NetworkStateType.UNKNOWN
+  }
+
+  private fun getConnectionType(netCapabilities: NetworkCapabilities?): NetworkStateType =
+    if (netCapabilities == null) NetworkStateType.UNKNOWN
```

Comment:
`null` case is handled separately in the beginning in order to avoid using safe calls `?.` later on